### PR TITLE
[Dialogue] Part 2: Translation from AtlasDB Services to ServiceConfigBlocks

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Allow us to, given Atlas level constructs for specifying how a service should be laid out (`ServerListConfig`s and `AuxiliaryRemotingParameters`-es, including having variable retry and blocking timeout behaviours), translate them into a suitable `ServiceConfigBlock` that Dialogue understands.

**Implementation Description (bullets)**:
- Extract the constants in `ClientOptions` to a separate class: these are used by both CJR and Dialogue. Refactor to `HumanReadableDuration` since converting from that to Java durations is easier than the other way around.
- Bump the version of Dialogue.
- Implement `DialogueClientOptions`, the conversion piece.

**Testing (What was existing testing like?  What have you done to improve it?)**:
New classes are tested. For the refactor, existing tests should cover it.

**Concerns (what feedback would you like?)**:
- This section is prone to typos (tests helped catch one). Were there any other silly mistakes?

**Where should we start reviewing?**: ClientOptionsConstants. It's probably worthwhile validating the constant extraction and the implementation of DialogueClientOptions as two parts.

**Priority (whenever / two weeks / yesterday)**: this week

### CODE REVIEW

File: atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/DialogueClientOptionsTest.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    private static final SslConfiguration SSL_CONFIGURATION = SslConfiguration.of(Paths.get("a", "b"));
+
+    private static final AuxiliaryRemotingParameters REMOTING_PARAMETERS_EXTENDED_TIMEOUT
+            = AuxiliaryRemotingParameters.builder()
+            .userAgent(USER_AGENT)
+            .shouldRetry(true)
+            .shouldLimitPayload(true)
+            .shouldUseExtendedTimeout(true)
+            .build();
+    private static final AuxiliaryRemotingParameters REMOTING_PARAMETERS = AuxiliaryRemotingParameters.builder()
```
Comment (gmaretic) [inline atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/DialogueClientOptionsTest.java:49]: name it _SHORT

File: atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/DialogueClientOptionsTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            .sslConfiguration(SSL_CONFIGURATION)
+            .addAllServers(SERVERS)
+            .build();
+
+    private static final RemoteServiceConfiguration REMOTE_SERVICE_CONFIGURATION_EXTENDED_TIMEOUT
+            = ImmutableRemoteServiceConfiguration.builder()
+            .serverList(SERVER_LIST_CONFIG)
+            .remotingParameters(REMOTING_PARAMETERS_EXTENDED_TIMEOUT)
+            .build();
+    private static final RemoteServiceConfiguration REMOTE_SERVICE_CONFIGURATION
```
Comment (gmaretic) [inline atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/DialogueClientOptionsTest.java:66]: Same as above

File: atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/DialogueClientOptionsTest.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+        PartialServiceConfiguration partialServiceConfiguration = servicesConfigBlock.services().get(SERVICE_NAME);
+        assertThat(partialServiceConfiguration.uris()).hasSameElementsAs(SERVERS);
+        assertThat(partialServiceConfiguration.security()).contains(SSL_CONFIGURATION);
+        assertThat(partialServiceConfiguration.backoffSlotSize())
+                .contains(ClientOptionsConstants.STANDARD_BACKOFF_SLOT_SIZE);
+        assertThat(partialServiceConfiguration.connectTimeout()).contains(ClientOptionsConstants.CONNECT_TIMEOUT);
+        assertThat(partialServiceConfiguration.maxNumRetries()).contains(ClientOptionsConstants.STANDARD_MAX_RETRIES);
+        assertThat(partialServiceConfiguration.readTimeout()).contains(ClientOptionsConstants.LONG_READ_TIMEOUT);
+    }
```
Comment (gmaretic) [inline atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/DialogueClientOptionsTest.java:87]: Verify absence of proxy configuration

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java [hunk 4]
```diff
@@ -82,7 +82,9 @@
 
     private static final int DEFAULT_LOCK_TIMEOUT_MS = 10_000;
     private static final int LONG_LOCK_TIMEOUT_MS =
```
Comment (gmaretic) [inline timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:84]: Maybe rename this

File: atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/DialogueClientOptionsTest.java (hunk 1 above)
Comment (jeremyk-91) [reply atlasdb-conjure/src/test/java/com/palantir/atlasdb/http/v2/DialogueClientOptionsTest.java:49]: Yep, makes sense - this sounds like a default.

File: timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java (hunk 4 above)
Comment (jeremyk-91) [reply timelock-server/src/suiteTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java:84]: Yep, renamed to `LONGER_THAN_READ_TIMEOUT_LOCK_TIMEOUT_MS`

File: atlasdb-conjure/build.gradle [hunk 5]
```diff
@@ -4,7 +4,12 @@ dependencies {
   compile project(':atlasdb-commons')
   compile project(':atlasdb-remoting-api')
 
-  compile group: 'com.palantir.conjure.java.runtime', name: 'conjure-java-jaxrs-client'
+  implementation 'com.palantir.refreshable:refreshable'
+  implementation 'com.palantir.dialogue:dialogue-apache-hc4-client'
```
Comment (iamdanfox) [inline atlasdb-conjure/build.gradle:8]: I think dialogue-clients should give you everything you need - don't think you need `-serde` or `-hc4-client`.

File: atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/DialogueClientOptions.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Map;
+
+import com.palantir.atlasdb.config.AuxiliaryRemotingParameters;
+import com.palantir.atlasdb.config.ServerListConfig;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.conjure.java.api.config.service.PartialServiceConfiguration;
+import com.palantir.conjure.java.api.config.service.ServicesConfigBlock;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+
+public final class DialogueClientOptions {
```
Comment (iamdanfox) [inline atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/DialogueClientOptions.java:28]: Can utilities like this be package-private? I know you guys don't use revapi right now, but I think if you made more of your stuff package-private maybe you could?

**Pull request merged**